Projects/b_u585i_iot02a_ntz/Src/ota_pal/ota_pal_stm32u5_ntz.c

```diff
@@ -222,103 +378,163 @@ static HAL_StatusTypeDef prvFlashSetDualBankMode( void )
-            HAL_FLASHEx_OBGetConfig( &OBInit );
+        /* Allow Access to Flash control registers and user Flash */
+        xHalStatus = HAL_FLASH_Unlock();
 
-            OBInit.OptionType = OPTIONBYTE_USER;
-            OBInit.USERType = OB_USER_SWAP_BANK;
+        if( xHalStatus != HAL_OK )
+        {
+            /* Lock on error */
+            ( void ) HAL_FLASH_Lock();
+        }
+        else /* xHalStatus == HAL_OK */
+        {
+            /* Allow Access to option bytes sector */
+            xHalStatus = HAL_FLASH_OB_Unlock();
 
-            if( ( ( OBInit.USERConfig ) & ( OB_SWAP_BANK_ENABLE ) ) != OB_SWAP_BANK_ENABLE )
+            if( xHalStatus != HAL_OK )
             {
-                OBInit.USERConfig = OB_SWAP_BANK_ENABLE;
-            }
-            else
-            {
-                OBInit.USERConfig = OB_SWAP_BANK_DISABLE;
+                /* Lock on error */
+                ( void ) HAL_FLASH_Lock();
             }
+        }
 
-            status = HAL_FLASHEx_OBProgram( &OBInit );
+        if( xHalStatus == HAL_OK )
+        {
+            /* Get the SWAP_BANK status */
+            HAL_FLASHEx_OBGetConfig( &xObContext );
 
-            if( status == HAL_OK )
+            xObContext.OptionType = OPTIONBYTE_USER;
+            xObContext.USERType = OB_USER_SWAP_BANK;
+
+            if( ulNewBank == FLASH_BANK_2 )
+            {
+                xObContext.USERConfig = OB_SWAP_BANK_ENABLE;
+            }
+            else /* ulNewBank == FLASH_BANK_1  */
             {
-                /* Generate System Reset to load the new option byte values ***************/
-                /* On successful option bytes loading the system should reset and control should not return from this function. */
-                status = HAL_FLASH_OB_Launch();
+                xObContext.USERConfig = OB_SWAP_BANK_DISABLE;
             }
 
+            xHalStatus = HAL_FLASHEx_OBProgram( &xObContext );
+
             ( void ) HAL_FLASH_OB_Lock();
+            ( void ) HAL_FLASH_Lock();
         }
 
-        ( void ) HAL_FLASH_Lock();
+        xResult = ( xHalStatus == HAL_OK ) ? pdTRUE : pdFALSE;
     }
 
-    return status;
+    return xResult;
 }
 
-
-static uint32_t prvGetActiveBank( void )
+static void prvOptionByteApply( void )
 {
-    HAL_StatusTypeDef status = HAL_ERROR;
-    uint32_t ulBank = 0UL;
+    vPetWatchdog();
 
-    status = HAL_FLASH_Unlock();
+    if( xTaskGetSchedulerState() == taskSCHEDULER_RUNNING )
+    {
+        vTaskSuspendAll();
+    }
 
-    if( status == HAL_OK )
+    LogSys( "Option Byte Launch in progress." );
+
+    /* Drain log buffers */
+    vDyingGasp();
+
+
+    ( void ) HAL_FLASH_Unlock();
+    ( void ) HAL_FLASH_OB_Unlock();
+    /* Generate System Reset to load the new option byte values ***************/
+    /* On successful option bytes loading the system should reset and control should not return from this function. */
+    ( void ) HAL_FLASH_OB_Launch();
```

Comment:
Do we lock back if the OB Launch fails so as to be in a consistent state as before.